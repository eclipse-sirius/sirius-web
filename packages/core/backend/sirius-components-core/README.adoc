= sirius-components-core

== Goal

This project is used to contain the API to manipulate core concepts that are not specific to the business of our application.

== Dependencies

- sirius-components-annotations

This project should not depend on anything else without a good reason to do so.