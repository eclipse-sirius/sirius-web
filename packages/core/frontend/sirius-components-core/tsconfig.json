{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@eclipse-sirius/sirius-components-tsconfig/react-library.json",
  "include": ["."],
  "exclude": ["dist", "build", "node_modules"]
}
