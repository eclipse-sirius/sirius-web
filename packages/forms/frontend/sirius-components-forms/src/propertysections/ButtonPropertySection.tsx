/*******************************************************************************
 * Copyright (c) 2022, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { useMutation } from '@apollo/client';
import { ServerContext, ServerContextValue, getCSSColor, useMultiToast } from '@eclipse-sirius/sirius-components-core';
import Button from '@material-ui/core/Button';
import { Theme, makeStyles } from '@material-ui/core/styles';
import gql from 'graphql-tag';
import { useContext, useEffect } from 'react';
import { PropertySectionComponent, PropertySectionComponentProps } from '../form/Form.types';
import { GQLButton } from '../form/FormEventFragments.types';
import {
  ButtonStyleProps,
  GQLErrorPayload,
  GQLPushButtonInput,
  GQLPushButtonMutationData,
  GQLPushButtonMutationVariables,
  GQLPushButtonPayload,
  GQLSuccessPayload,
} from './ButtonPropertySection.types';
import { PropertySectionLabel } from './PropertySectionLabel';
import { getTextDecorationLineValue } from './getTextDecorationLineValue';

const useStyle = makeStyles<Theme, ButtonStyleProps>((theme) => ({
  style: {
    backgroundColor: ({ backgroundColor }) =>
      backgroundColor ? getCSSColor(backgroundColor, theme) : theme.palette.primary.light,
    color: ({ foregroundColor }) => (foregroundColor ? getCSSColor(foregroundColor, theme) : 'white'),
    fontSize: ({ fontSize }) => (fontSize ? fontSize : null),
    fontStyle: ({ italic }) => (italic ? 'italic' : null),
    fontWeight: ({ bold }) => (bold ? 'bold' : null),
    textDecorationLine: ({ underline, strikeThrough }) => getTextDecorationLineValue(underline, strikeThrough),
    paddingTop: theme.spacing(0.5),
    paddingBottom: theme.spacing(0.5),
    '&:hover': {
      backgroundColor: ({ backgroundColor }) =>
        backgroundColor ? getCSSColor(backgroundColor, theme) : theme.palette.primary.main,
      color: ({ foregroundColor }) => (foregroundColor ? getCSSColor(foregroundColor, theme) : 'white'),
      fontSize: ({ fontSize }) => (fontSize ? fontSize : null),
      fontStyle: ({ italic }) => (italic ? 'italic' : null),
      fontWeight: ({ bold }) => (bold ? 'bold' : null),
      textDecorationLine: ({ underline, strikeThrough }) => getTextDecorationLineValue(underline, strikeThrough),
    },
  },
  icon: {
    marginRight: ({ iconOnly }) => (iconOnly ? theme.spacing(0) : theme.spacing(2)),
  },
  propertySection: {
    display: 'flex',
    flexDirection: 'row',
    gap: theme.spacing(2),
  },
}));

export const pushButtonMutation = gql`
  mutation pushButton($input: PushButtonInput!) {
    pushButton(input: $input) {
      __typename
      ... on ErrorPayload {
        messages {
          body
          level
        }
      }
      ... on SuccessPayload {
        messages {
          body
          level
        }
      }
    }
  }
`;

const isErrorPayload = (payload: GQLPushButtonPayload): payload is GQLErrorPayload =>
  payload.__typename === 'ErrorPayload';

const isSuccessPayload = (payload: GQLPushButtonPayload): payload is GQLSuccessPayload =>
  payload.__typename === 'SuccessPayload';

/**
 * Defines the content of a Button property section.
 * The content is submitted when pushing the button.
 */
export const ButtonPropertySection: PropertySectionComponent<GQLButton> = ({
  editingContextId,
  formId,
  widget,
  readOnly,
}: PropertySectionComponentProps<GQLButton>) => {
  const props: ButtonStyleProps = {
    backgroundColor: widget.style?.backgroundColor ?? null,
    foregroundColor: widget.style?.foregroundColor ?? null,
    fontSize: widget.style?.fontSize ?? null,
    italic: widget.style?.italic ?? null,
    bold: widget.style?.bold ?? null,
    underline: widget.style?.underline ?? null,
    strikeThrough: widget.style?.strikeThrough ?? null,
    iconOnly: widget.buttonLabel ? false : true,
  };
  const classes = useStyle(props);

  const { httpOrigin } = useContext<ServerContextValue>(ServerContext);

  const [pushButton, { loading, data, error }] = useMutation<GQLPushButtonMutationData, GQLPushButtonMutationVariables>(
    pushButtonMutation
  );

  const { addErrorMessage, addMessages } = useMultiToast();

  useEffect(() => {
    if (!loading) {
      if (error) {
        addErrorMessage('An unexpected error has occurred, please refresh the page');
      }
      if (data) {
        const { pushButton } = data;
        if (isErrorPayload(pushButton) || isSuccessPayload(pushButton)) {
          addMessages(pushButton.messages);
        }
      }
    }
  }, [loading, error, data]);

  const onClick = () => {
    const input: GQLPushButtonInput = {
      id: crypto.randomUUID(),
      editingContextId,
      representationId: formId,
      buttonId: widget.id,
    };
    const variables: GQLPushButtonMutationVariables = { input };
    pushButton({ variables });
  };

  return (
    <div className={classes.propertySection}>
      <PropertySectionLabel
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        data-testid={widget.label}
      />
      <Button
        data-testid={widget.buttonLabel}
        variant="contained"
        color="primary"
        onClick={onClick}
        disabled={readOnly || widget.readOnly}
        classes={{ root: classes.style }}>
        {widget.imageURL?.length > 0 ? (
          <img className={classes.icon} width="16" height="16" alt={widget.label} src={httpOrigin + widget.imageURL} />
        ) : null}
        {widget.buttonLabel}
      </Button>
    </div>
  );
};
