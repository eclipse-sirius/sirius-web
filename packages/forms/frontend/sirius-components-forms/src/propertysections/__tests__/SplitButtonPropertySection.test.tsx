/*******************************************************************************
 * Copyright (c) 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { MessageOptions, ServerContext, ToastContext, ToastContextValue } from '@eclipse-sirius/sirius-components-core';
import { act, cleanup, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { afterEach, expect, test, vi } from 'vitest';
import { GQLButton, GQLSplitButton } from '../../form/FormEventFragments.types';
import { pushButtonMutation } from '../ButtonPropertySection';
import {
  GQLErrorPayload,
  GQLPushButtonMutationData,
  GQLPushButtonMutationVariables,
  GQLSuccessPayload,
} from '../ButtonPropertySection.types';
import { SplitButtonPropertySection } from '../SplitButtonPropertySection';

crypto.randomUUID = vi.fn(() => '48be95fc-3422-45d3-b1f9-d590e847e9e1');

afterEach(() => cleanup());

const defaultButton: GQLButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: [],
  hasHelpText: false,
  diagnostics: [],
  buttonLabel: 'ButtonLabel',
  imageURL: null,
  style: {
    backgroundColor: null,
    foregroundColor: null,
    fontSize: null,
    italic: null,
    bold: null,
    underline: null,
    strikeThrough: null,
  },
  readOnly: false,
};

const defaultSplitButton: GQLSplitButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: [],
  hasHelpText: false,
  diagnostics: [],
  readOnly: false,
  actions: [defaultButton, defaultButton],
};

const buttonWithStyle: GQLButton = {
  ...defaultButton,
  style: {
    backgroundColor: '#de1000',
    foregroundColor: '#fbb800',
    fontSize: 14,
    italic: false,
    bold: false,
    underline: false,
    strikeThrough: false,
  },
};

const readOnlyButton: GQLButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: [],
  diagnostics: [],
  buttonLabel: 'ButtonLabel',
  imageURL: null,
  style: {
    backgroundColor: null,
    foregroundColor: null,
    fontSize: null,
    italic: null,
    bold: null,
    underline: null,
    strikeThrough: null,
  },
  readOnly: true,
  hasHelpText: false,
};

const splitButtonWithStyle: GQLSplitButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: [],
  hasHelpText: false,
  diagnostics: [],
  readOnly: false,
  actions: [buttonWithStyle, buttonWithStyle],
};

const splitButtonWithReadOnlyAction: GQLSplitButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: [],
  hasHelpText: false,
  diagnostics: [],
  readOnly: false,
  actions: [readOnlyButton, readOnlyButton],
};

const pushButtonVariables: GQLPushButtonMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    buttonId: 'buttonId',
  },
};
const successPayload: GQLSuccessPayload = { __typename: 'SuccessPayload', messages: [] };
const pushButtonSuccessData: GQLPushButtonMutationData = { pushButton: successPayload };

const pushButtonErrorPayload: GQLErrorPayload = {
  __typename: 'ErrorPayload',
  messages: [
    {
      body: 'An error has occurred, please refresh the page',
      level: 'ERROR',
    },
    {
      body: 'FeedbackMessage',
      level: 'INFO',
    },
  ],
};
const pushButtonErrorData: GQLPushButtonMutationData = {
  pushButton: pushButtonErrorPayload,
};

const mockEnqueue = vi.fn<[string, MessageOptions?], void>();

const toastContextMock: ToastContextValue = {
  enqueueSnackbar: mockEnqueue,
};

test('should render the split button and his actions', () => {
  const { container } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={defaultSplitButton}
            readOnly={false}
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  screen.getByRole('show-actions').click();
  expect(container).toMatchSnapshot();
});

test('should render a readOnly button and his actions', () => {
  const { container } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={defaultSplitButton}
            readOnly
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  screen.getByRole('show-actions').click();
  expect(container).toMatchSnapshot();
});

test('should send mutation when clicked', async () => {
  let pushButtonCalled = false;
  const pushButtonSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: pushButtonMutation,
      variables: pushButtonVariables,
    },
    result: () => {
      pushButtonCalled = true;
      return { data: pushButtonSuccessData };
    },
  };

  const mocks = [pushButtonSuccessMock];
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={defaultSplitButton}
            readOnly={false}
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();

  const element: HTMLElement = screen.getByTestId('Label');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(pushButtonCalled).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });
});

test('should display the error received', async () => {
  let pushButtonCalled = false;
  const pushButtonErrorMock: MockedResponse<Record<string, any>> = {
    request: {
      query: pushButtonMutation,
      variables: pushButtonVariables,
    },
    result: () => {
      pushButtonCalled = true;
      return { data: pushButtonErrorData };
    },
  };

  const mocks = [pushButtonErrorMock];
  const { baseElement } = render(
    <MockedProvider mocks={mocks}>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={defaultSplitButton}
            readOnly={false}
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();

  const element: HTMLElement = screen.getByTestId('Label');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(pushButtonCalled).toBeTruthy();
      expect(mockEnqueue).toHaveBeenCalledTimes(2);
      expect(baseElement).toMatchSnapshot();
    });
  });
});

test('should render the button and his actions with style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={splitButtonWithStyle}
            readOnly={false}
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  screen.getByRole('show-actions').click();
  expect(baseElement).toMatchSnapshot();
});

test('should render the button with help hint', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        {' '}
        <ToastContext.Provider value={toastContextMock}>
          <SplitButtonPropertySection
            editingContextId="editingContextId"
            formId="formId"
            widget={{ ...defaultSplitButton, hasHelpText: true }}
            readOnly={false}
          />
        </ToastContext.Provider>
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});
