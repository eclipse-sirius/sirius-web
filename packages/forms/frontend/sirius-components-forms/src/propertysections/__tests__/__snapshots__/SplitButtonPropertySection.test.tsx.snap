// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should display the error received 1`] = `
<body>
  <div>
    <div
      class="makeStyles-style-40"
    >
      <div
        class="makeStyles-propertySectionLabel-41"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-44 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <div
        aria-label="split button"
        class="MuiButtonGroup-root MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-style-34 makeStyles-style-36 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
          data-testid="Label"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            ButtonLabel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-haspopup="menu"
          aria-label="select button action"
          class="MuiButtonBase-root MuiButton-root makeStyles-style-34 makeStyles-style-36 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          role="show-actions"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`should display the error received 2`] = `
<body>
  <div>
    <div
      class="makeStyles-style-40"
    >
      <div
        class="makeStyles-propertySectionLabel-41"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-44 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <div
        aria-label="split button"
        class="MuiButtonGroup-root MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-style-34 makeStyles-style-36 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
          data-testid="Label"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            ButtonLabel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-haspopup="menu"
          aria-label="select button action"
          class="MuiButtonBase-root MuiButton-root makeStyles-style-34 makeStyles-style-36 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          role="show-actions"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render a readOnly button and his actions 1`] = `
<div>
  <div
    class="makeStyles-style-18"
  >
    <div
      class="makeStyles-propertySectionLabel-19"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-22 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <div
      aria-label="split button"
      class="MuiButtonGroup-root MuiButtonGroup-contained"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-12 makeStyles-style-14 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="Label"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
      </button>
      <button
        aria-haspopup="menu"
        aria-label="select button action"
        class="MuiButtonBase-root MuiButton-root makeStyles-style-12 makeStyles-style-14 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
        disabled=""
        role="show-actions"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`should render the button and his actions with style 1`] = `
<body>
  <div>
    <div
      class="makeStyles-style-51"
    >
      <div
        class="makeStyles-propertySectionLabel-52"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-55 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <div
        aria-label="split button"
        class="MuiButtonGroup-root MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-style-45 makeStyles-style-47 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
          data-testid="Label"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            ButtonLabel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-controls="split-button-menu"
          aria-expanded="true"
          aria-haspopup="menu"
          aria-label="select button action"
          class="MuiButtonBase-root MuiButton-root makeStyles-style-45 makeStyles-style-47 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          role="show-actions"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    role="tooltip"
    style="position: absolute; top: 0px; left: 0px;"
    x-placement="bottom"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      style="opacity: 1; transform: scale(1, 1); transform-origin: center top; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <ul
        class="MuiList-root MuiList-padding"
        id="split-button-menu"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root makeStyles-style-45 makeStyles-style-47 Mui-selected MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button Mui-selected"
          role="menuitem"
          tabindex="0"
        >
          ButtonLabel
          <span
            class="MuiTouchRipple-root"
          />
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root makeStyles-style-45 makeStyles-style-49 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          role="menuitem"
          tabindex="-1"
        >
          ButtonLabel
          <span
            class="MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`should render the button with help hint 1`] = `
<body>
  <div>
     
    <div
      class="makeStyles-style-62"
    >
      <div
        class="makeStyles-propertySectionLabel-63"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-66 MuiTypography-subtitle2"
        >
          Label
        </h6>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
          focusable="false"
          style="margin-left: 8px; font-size: 16px;"
          title=""
          viewBox="0 0 24 24"
        >
          <path
            d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
          />
        </svg>
      </div>
      <div
        aria-label="split button"
        class="MuiButtonGroup-root MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-style-56 makeStyles-style-58 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
          data-testid="Label"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            ButtonLabel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-haspopup="menu"
          aria-label="select button action"
          class="MuiButtonBase-root MuiButton-root makeStyles-style-56 makeStyles-style-58 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          role="show-actions"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render the split button and his actions 1`] = `
<div>
  <div
    class="makeStyles-style-7"
  >
    <div
      class="makeStyles-propertySectionLabel-8"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-11 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <div
      aria-label="split button"
      class="MuiButtonGroup-root MuiButtonGroup-contained"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-1 makeStyles-style-3 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
        data-testid="Label"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-controls="split-button-menu"
        aria-expanded="true"
        aria-haspopup="menu"
        aria-label="select button action"
        class="MuiButtonBase-root MuiButton-root makeStyles-style-1 makeStyles-style-3 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        role="show-actions"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`should send mutation when clicked 1`] = `
<div>
  <div
    class="makeStyles-style-29"
  >
    <div
      class="makeStyles-propertySectionLabel-30"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-33 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <div
      aria-label="split button"
      class="MuiButtonGroup-root MuiButtonGroup-contained"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-23 makeStyles-style-25 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
        data-testid="Label"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-haspopup="menu"
        aria-label="select button action"
        class="MuiButtonBase-root MuiButton-root makeStyles-style-23 makeStyles-style-25 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        role="show-actions"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`should send mutation when clicked 2`] = `
<div>
  <div
    class="makeStyles-style-29"
  >
    <div
      class="makeStyles-propertySectionLabel-30"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-33 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <div
      aria-label="split button"
      class="MuiButtonGroup-root MuiButtonGroup-contained"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-23 makeStyles-style-25 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary"
        data-testid="Label"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-haspopup="menu"
        aria-label="select button action"
        class="MuiButtonBase-root MuiButton-root makeStyles-style-23 makeStyles-style-25 MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        role="show-actions"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
